File: python-bridge/start-bridge.js
/**
 * JobSpy Python Bridge Starter
 * 
 * This script starts the JobSpy Python bridge service and handles errors.
 */

const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');
const os = require('os');

// Configuration
const BRIDGE_HOST = process.env.JOBSPY_BRIDGE_HOST || '127.0.0.1'; // Force IPv4
const BRIDGE_PORT = process.env.JOBSPY_BRIDGE_PORT || 8000;
const PID_FILE = path.join(__dirname, 'bridge_pid.txt');
const LOG_FILE = path.join(__dirname, 'logs', 'bridge.log');

// Ensure logs directory exists
if (!fs.existsSync(path.join(__dirname, 'logs'))) {
  fs.mkdirSync(path.join(__dirname, 'logs'), { recursive: true });
}

// Simple logger that writes to console and file
const logger = {
  info: (message) => {
    const logMessage = `[INFO] ${new Date().toISOString()} - ${message}`;
    console.log(logMessage);
    fs.appendFileSync(LOG_FILE, logMessage + '\n');
  },
  error: (message) => {
    const logMessage = `[ERROR] ${new Date().toISOString()} - ${message}`;
    console.error(logMessage);
    fs.appendFileSync(LOG_FILE, logMessage + '\n');
  },
  warn: (message) => {
    const logMessage = `[WARN] ${new Date().toISOString()} - ${message}`;
    console.warn(logMessage);
    fs.appendFileSync(LOG_FILE, logMessage + '\n');
  }
};

// Check if bridge is already running
function isBridgeRunning() {
  if (fs.existsSync(PID_FILE)) {
    try {
      const pid = parseInt(fs.readFileSync(PID_FILE, 'utf8'));
      
      // Check if process is running
      try {
        process.kill(pid, 0); // This doesn't actually kill the process, just checks if it exists
        return true;
      } catch (e) {
        // Process doesn't exist, clean up PID file
        fs.unlinkSync(PID_FILE);
        return false;
      }
    } catch (error) {
      logger.warn(`Error checking bridge PID: ${error.message}`);
      return false;
    }
  }
  
  return false;
}

// Kill any existing bridge process
function killExistingBridge() {
  if (fs.existsSync(PID_FILE)) {
    try {
      const pid = parseInt(fs.readFileSync(PID_FILE, 'utf8'));
      
      // Try to kill the process
      try {
        process.kill(pid);
        logger.info(`Killed existing bridge process with PID ${pid}`);
      } catch (e) {
        logger.warn(`Failed to kill process ${pid}: ${e.message}`);
      }
      
      // Remove PID file
      fs.unlinkSync(PID_FILE);
    } catch (error) {
      logger.error(`Error killing existing bridge: ${error.message}`);
    }
  }
}

// Start the bridge
function startBridge() {
  try {
    // Check if bridge is already running
    if (isBridgeRunning()) {
      logger.info('Bridge is already running');
      return;
    }
    
    // Find Python executable
    const isPythonInstalled = (cmd) => {
      try {
        require('child_process').execSync(`${cmd} --version`, { stdio: 'ignore' });
        return true;
      } catch (e) {
        return false;
      }
    };
    
    let pythonCmd = 'python3';
    if (!isPythonInstalled(pythonCmd)) {
      pythonCmd = 'python';
      if (!isPythonInstalled(pythonCmd)) {
        throw new Error('Python not found. Please install Python 3.');
      }
    }
    
    logger.info(`Starting JobSpy bridge on ${BRIDGE_HOST}:${BRIDGE_PORT}`);
    
    // Determine platform-specific command
    const args = [
      '-m', 'uvicorn',
      'jobspy_bridge:app',
      '--host', BRIDGE_HOST,
      '--port', BRIDGE_PORT.toString()
    ];
    
    const options = {
      cwd: __dirname,
      detached: true,
      stdio: ['ignore', 'pipe', 'pipe'],
      env: {
        ...process.env,
        PYTHONUNBUFFERED: '1', // Force unbuffered output
        JOBSPY_BRIDGE_HOST: BRIDGE_HOST,
        JOBSPY_BRIDGE_PORT: BRIDGE_PORT.toString()
      }
    };
    
    // Start the bridge process
    const bridgeProcess = spawn(pythonCmd, args, options);
    
    // Write PID to file
    fs.writeFileSync(PID_FILE, bridgeProcess.pid.toString());
    
    logger.info(`Started JobSpy bridge with PID ${bridgeProcess.pid}`);
    
    // Handle process events
    bridgeProcess.stdout.on('data', (data) => {
      logger.info(`Bridge output: ${data.toString().trim()}`);
    });
    
    bridgeProcess.stderr.on('data', (data) => {
      const message = data.toString().trim();
      if (message.includes('ERROR')) {
        logger.error(`Bridge error: ${message}`);
      } else {
        logger.info(`Bridge log: ${message}`);
      }
    });
    
    bridgeProcess.on('error', (error) => {
      logger.error(`Failed to start bridge: ${error.message}`);
    });
    
    bridgeProcess.on('close', (code) => {
      logger.warn(`Bridge process exited with code ${code}`);
      // Remove PID file when process exits
      if (fs.existsSync(PID_FILE)) {
        fs.unlinkSync(PID_FILE);
      }
    });
    
    // Detach process so it stays running when this script exits
    bridgeProcess.unref();
    
  } catch (error) {
    logger.error(`Error starting bridge: ${error.message}`);
    throw error;
  }
}

// Main function
function main() {
  try {
    logger.info('Starting JobSpy bridge service');
    killExistingBridge(); // Always kill existing bridge to ensure clean startup
    startBridge();
  } catch (error) {
    logger.error(`Failed to start bridge: ${error.message}`);
    process.exit(1);
  }
}

// Run main function
main(); 

File: python-bridge/start-bridge.py
#!/usr/bin/env python3
import os
import sys
import subprocess
import time
import logging
import signal
import socket
from pathlib import Path

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.StreamHandler(),
        logging.FileHandler("start-bridge.log")
    ]
)
logger = logging.getLogger("bridge-starter")

# Ensure the bridge process can be killed properly
bridge_process = None

def signal_handler(sig, frame):
    logger.info(f"Received signal {sig}, shutting down bridge...")
    if bridge_process:
        try:
            bridge_process.terminate()
            logger.info("Bridge process terminated")
        except Exception as e:
            logger.error(f"Error terminating bridge process: {e}")
    sys.exit(0)

# Register signal handlers
signal.signal(signal.SIGINT, signal_handler)
signal.signal(signal.SIGTERM, signal_handler)

def wait_for_port(host, port, timeout=30):
    """Wait for a port to be available on a host."""
    start_time = time.time()
    while True:
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
                sock.settimeout(1)
                result = sock.connect_ex((host, port))
                if result == 0:
                    logger.info(f"Port {port} is open on {host}")
                    return True
        except Exception as e:
            logger.debug(f"Error checking port: {e}")
        
        if time.time() - start_time > timeout:
            logger.error(f"Timeout waiting for port {port} on {host}")
            return False
        
        logger.info(f"Waiting for port {port} to be available (elapsed: {int(time.time() - start_time)}s)...")
        time.sleep(2)

def main():
    global bridge_process
    
    # Get the directory of this script
    script_dir = Path(__file__).parent.absolute()
    project_root = script_dir.parent
    
    # Path to the bridge script
    bridge_script = project_root / "jobspy_bridge.py"
    
    if not bridge_script.exists():
        logger.error(f"Bridge script not found at {bridge_script}")
        sys.exit(1)
    
    # Virtual environment path
    venv_path = project_root / ".venv"
    if venv_path.exists():
        # Use Python from virtual environment
        python_executable = venv_path / "bin" / "python"
        # Check if the path exists before using it
        if not python_executable.exists():
            # Try Windows path
            python_executable = venv_path / "Scripts" / "python.exe"
            if not python_executable.exists():
                logger.warning(f"Python executable not found in venv, falling back to system Python")
                python_executable = "python"
    else:
        # Use system Python
        python_executable = "python"
    
    # Ensure network settings are configured for IPv4
    logger.info("Setting up IPv4 network configuration...")
    try:
        ipv4_addr = socket.gethostbyname("localhost")
        logger.info(f"IPv4 address for localhost: {ipv4_addr}")
    except Exception as e:
        logger.warning(f"Could not resolve localhost: {e}")
    
    # Create a logs directory if it doesn't exist
    logs_dir = project_root / "logs"
    logs_dir.mkdir(exist_ok=True)
    
    # Path for bridge output log
    bridge_log = logs_dir / "jobspy_bridge.log"
    
    logger.info(f"Starting bridge from {bridge_script} with Python: {python_executable}")
    
    try:
        # Start the bridge process
        bridge_output = open(bridge_log, "w")
        bridge_process = subprocess.Popen(
            [python_executable, str(bridge_script)],
            stdout=bridge_output,
            stderr=subprocess.STDOUT,
            cwd=str(project_root)
        )
        
        logger.info(f"Started bridge process with PID: {bridge_process.pid}")
        
        # Wait for the bridge to start
        bridge_host = os.environ.get("JOBSPY_BRIDGE_HOST", "127.0.0.1")
        bridge_port = int(os.environ.get("JOBSPY_BRIDGE_PORT", "8000"))
        
        logger.info(f"Waiting for bridge to be available on {bridge_host}:{bridge_port}...")
        if wait_for_port(bridge_host, bridge_port):
            logger.info("Bridge is now available")
            
            # Keep the process running
            logger.info("Bridge is running. Press Ctrl+C to stop.")
            bridge_process.wait()
        else:
            logger.error("Bridge failed to start")
            if bridge_process.poll() is not None:
                logger.error(f"Bridge process exited with code: {bridge_process.returncode}")
            
            # Output the bridge log for debugging
            logger.error("Bridge log output:")
            with open(bridge_log, "r") as log_file:
                logger.error(log_file.read())
            
            sys.exit(1)
    except Exception as e:
        logger.error(f"Error starting bridge: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main() 